Mathlib/LinearAlgebra/Matrix/SpecialLinearGroup.lean

```diff
@@ -251,6 +259,81 @@ def map (f : R →+* S) : SpecialLinearGroup n R →* SpecialLinearGroup n S whe
   map_mul' x y := Subtype.ext <| f.mapMatrix.map_mul ↑ₘx ↑ₘy
 #align matrix.special_linear_group.map Matrix.SpecialLinearGroup.map
 
+section center
+
+open Subgroup
+
+@[simp]
+theorem center_eq_bot_of_subsingleton [Subsingleton n] :
+    center (SpecialLinearGroup n R) = ⊥ :=
+  eq_bot_iff.mpr fun x _ ↦ by rw [mem_bot, Subsingleton.elim x 1]
+
+theorem scalar_eq_self_of_mem_center
+    {A : SpecialLinearGroup n R} (hA : A ∈ center (SpecialLinearGroup n R)) (i : n) :
+    scalar n (A i i) = A := by
+  obtain ⟨r : R, hr : scalar n r = A⟩ := mem_range_scalar_of_commute_transvectionStruct fun t ↦
+    Subtype.ext_iff.mp <| Subgroup.mem_center_iff.mp hA ⟨t.toMatrix, by simp⟩
+  simp [← congr_fun₂ hr i i, ← hr]
+
+theorem scalar_eq_coe_self_center
+    (A : center (SpecialLinearGroup n R)) (i : n) :
+    scalar n ((A : Matrix n n R) i i) = A :=
+  scalar_eq_self_of_mem_center A.property i
+
+/-- The center of a special linear group of degree `n` is the subgroup of scalar matrices, for which
+the scalars are the `n`-th roots of unity.-/
+theorem mem_center_iff {A : SpecialLinearGroup n R} :
+    A ∈ center (SpecialLinearGroup n R) ↔ ∃ (r : R), r ^ (Fintype.card n) = 1 ∧ scalar n r = A := by
+  rcases isEmpty_or_nonempty n with hn | ⟨⟨i⟩⟩; · exact ⟨by aesop, by simp [Subsingleton.elim A 1]⟩
+  refine ⟨fun h ↦  ⟨A i i, ?_, ?_⟩, fun ⟨r, _, hr⟩ ↦ mem_center_iff.mpr fun B ↦ ?_⟩
+  · have : det ((scalar n) (A i i)) = 1 := (scalar_eq_self_of_mem_center h i).symm ▸ A.property
+    simpa using this
+  · exact scalar_eq_self_of_mem_center h i
+  · suffices ↑ₘ(B * A) = ↑ₘ(A * B) from Subtype.val_injective this
+    simpa only [coe_mul, ← hr] using (scalar_commute (n := n) r (Commute.all r) B).symm
+
+/-- An equivalence of groups, from the center of the special linear group to the roots of unity. -/
+def center_equiv_rootsOfUnity' (i : n) :
+    center (SpecialLinearGroup n R) ≃* rootsOfUnity (Fintype.card n).toPNat' R where
+  toFun A := rootsOfUnity.mkOfPowEq (↑ₘA i i) <| by
+    have : Nonempty n := ⟨i⟩
+    obtain ⟨r, hr, hr'⟩ := mem_center_iff.mp A.property
+    replace hr' : A.val i i = r := by simp [← hr']
+    simp [hr, hr']
+  invFun a := ⟨⟨a • (1 : Matrix n n R), by aesop⟩,
+    Subgroup.mem_center_iff.mpr fun B ↦ Subtype.val_injective <| by simp [coe_mul]⟩
+  left_inv A := by
+    refine SetCoe.ext $ SetCoe.ext ?_
+    obtain ⟨r, _, hr⟩ := mem_center_iff.mp A.property
+    simp only [instHSMul, SMul.smul, Submonoid.coe_subtype, ← hr, scalar_apply, diagonal, of_apply,
+      ↓reduceIte, SMul.comp.smul]
+    aesop
```

Comment:
You should never unfold instance definitions (in this case you're doing this by including `instHSMul` and `SMul.smul` in your list). Instead you should use API lemmas (and add them if they are missing).

In this case there is no missing API so this proof can be just:
```suggestion
    simpa [← hr, Submonoid.smul_def, Units.smul_def] using smul_one_eq_diagonal r
```

Reply:
An advantage of this approach is the proofs like this one are not sensitive to the instance definition, but depend merely upon its API (making future refactoring much easier).